ddtrace/tracer/sampler.go

```diff
@@ -147,3 +154,177 @@ func (ps *prioritySampler) apply(spn *span) {
+		rules = rules[:0]
+		err := json.Unmarshal([]byte(rulesFromEnv), &rules)
+		if err != nil {
+			log.Warn("error parsing DD_TRACE_SAMPLING_RULES: %v", err)
+			return nil
+		}
+	}
+	validRules := make([]samplingRule, 0, len(rules))
+	for _, v := range rules {
+		s, err := regexp.Compile(v.Service)
+		if err != nil {
+			log.Warn("ignoring rule %v: %v", v, err)
+		}
+		n, err := regexp.Compile(v.Name)
+		if err != nil {
+			log.Warn("ignoring rule %v: %v", v, err)
+		}
+		if v.Rate < 0.0 || v.Rate > 1.0 {
+			log.Warn("ignoring rule %v: rate is out of range", v)
+		}
+		validRules = append(validRules, samplingRule{
+			service: s,
+			name:    n,
+			rate:    v.Rate,
+		})
+	}
+	return validRules
+}
+
+func sampleRate() float64 {
+	rate := 1.0
+	v := os.Getenv("DD_TRACE_SAMPLE_RATE")
+	if v == "" {
+		return rate
+	}
+	r, err := strconv.ParseFloat(v, 64)
+	if err != nil {
+		log.Warn("using default rate %f because DD_TRACE_SAMPLE_RATE is invalid: %v", rate, err)
+		return rate
+	}
+	if r >= 0.0 && r <= 1.0 {
+		return r
+	}
+	log.Warn("using default rate %f because provided value is out of range: %f", rate, r)
+	return rate
+}
+
+func rateLimiter(r float64) *rate.Limiter {
+	v := os.Getenv("DD_TRACE_RATE_LIMIT")
+	if v == "" {
+		return nil
+	}
+	l, err := strconv.ParseFloat(v, 64)
+	if err != nil {
+		log.Warn("using default rate limit because DD_TRACE_RATE_LIMIT is invalid: %v", err)
+		return nil
+	}
+	switch {
+	case l < 0.0:
+		return nil
+	case l == 0.0:
+		return rate.NewLimiter(0.0, 0)
+	case (r * l) < 1.0:
+		return rate.NewLimiter(rate.Limit(l), 1)
+	default:
+		return rate.NewLimiter(rate.Limit(l), int(math.Ceil(r*l)))
+	}
+}
+
+func (rs *rulesSampler) apply(span *span) bool {
+	matched := false
+	sr := 0.0
+	for _, v := range rs.rules {
+		if v.match(span) {
+			matched = true
+			sr = v.rate
+			break
+		}
+	}
+	if !matched {
+		return false
+	}
+	// rate sample
+	span.SetTag("_dd.rule_psr", sr)
+	if !sampledByRate(span.TraceID, sr) {
+		span.SetTag(ext.SamplingPriority, ext.PriorityAutoReject)
+		return true
+	}
+	// global rate limit and effective rate calculations
+	defer rs.mu.Unlock()
+	rs.mu.Lock()
+	if ts := time.Now().Unix(); ts > rs.ts {
+		// update "previous rate" and reset
+		if ts-rs.ts == 1 && rs.total > 0 && rs.allowed > 0 {
+			rs.previousRate = float64(rs.allowed) / float64(rs.total)
+		} else {
+			rs.previousRate = 0.0
+		}
+		rs.ts = ts
+		rs.allowed = 0
+		rs.total = 0
+	}
+
+	rs.total++
+	// calculate effective rate, and tag the span
+	er := (rs.previousRate + (float64(rs.allowed) / float64(rs.total))) / 2.0
+	span.SetTag("_dd.limit_psr", er)
+	if !rs.limiter.Allow() {
+		span.SetTag(ext.SamplingPriority, ext.PriorityAutoReject)
+		return true
+	}
+	span.SetTag(ext.SamplingPriority, ext.PriorityAutoKeep)
+	rs.allowed++
+
+	return true
+}
+
+// SamplingRule placeholder comment
+type SamplingRule struct {
+	Service string  `json:"service"`
+	Name    string  `json:"name"`
+	Rate    float64 `json:"rate"`
```

Comment:
Would it be better to just reuse this struct and have some unexported fields containing the compiled regex?

Reply:
For now, it doesn't seem like it - it's an input to configure sampling rules, but doesn't necessarily decide what is internally used as those rules.
It could be changed in the future without really affecting public API.

Reply:
I'd prefer if we used this struct instead of having another one and just have two more fields, something like `regexpService` and `regexpName`. It'd be much easier to reason about and manage.

Reply:
Resolving. We discussed in private keeping both is better.